### PR TITLE
fixed:issue->#1835 Support NYNAB import of transactions that contain subtransactions tha…

### PR DESCRIPTION
**Pull request by Marethyu1**
…t are transfers

A subtransaction is considered a transfer from ynab if
- the transfer_account_id is set
- the category_id is null (transfers dont have categories)

This addresses #1835

### CODE REVIEW

**Comment by MatissJanis** [comment]
Good catch. I like this fix, but I'm a bit worried about merging it as it exposes a different issue: sub-transaction transfers do not appear correctly.

Reproduction:
1. use the demo budget in the issue ([ynab5-demo-budget.json](https://github.com/actualbudget/actual/files/13257078/ynab5-demo-budget.json))
2. import it
3. observe "200" in the "Saving" account
4. expectation: the account has 250

So basically 50$ just disappeared from the budget. That's definitely not something we want.

I wonder if we should fix this by disallowing transfers in sub-transactions. 🤔

**Comment by Marethyu1** [comment]
I suspect that this is a mistake on my part - I'll take a look. The reason I believe that is
- I would also expect the savings account to have 250
- I updated the onboarding test to expect 250 in the savings account and it passes (this is what suprises me)

**Comment by Marethyu1** [comment]
Ah so, the initial test budget I uploaded is wrong. I have updated the issue with a correct budget, and also you can check It out here [ynab5-demo-budget.json](https://github.com/actualbudget/actual/files/13257675/ynab5-demo-budget.json)

The problem with the original budget I uploaded was that it was missing the corresponding transaction in savings account of +50 coming into the savings account. 

The reason why the tests were passing on this pr was that I fixed the issue in the pull request but didn't go back and update the linked budget in #1835 

TLDR: The changes in this pr should now match the new demo budget in #1835. These should now result in the savings account to have 250

File: packages/loot-core/src/server/importers/ynab5.ts [hunk 1]
```diff
@@ -178,11 +178,21 @@ async function importTransactions(
...
             subtransactions: subtransactions
               ? subtransactions.map(subtrans => {
+                  let payee = null;
+                  if (subtrans.transfer_account_id) {
+                    payee = payees.find(
+                      p =>
+                        p.transfer_acct ===
+                        entityIdMap.get(subtrans.transfer_account_id),
+                    ).id;
+                  }
```
Comment (MatissJanis) [inline packages/loot-core/src/server/importers/ynab5.ts:188]: **💬 suggestion:** ‏I suspect for bigger imports this operation will have poor performance. Would you mind pre-computing a hashmap for payees and then just using it here as a lookup?

i.e.
```
const payeeTransferActHashMap = new Map(payees.map((payee) => [payee.transfer_acct, payee]));

// ... and then inside the loop
payee = payeeTransferActHashMap[entityIdMap.get(subtrans.transfer_account_id)]?.id;

File: packages/loot-core/src/server/importers/ynab5.ts [hunk 2]
```diff
@@ -178,11 +178,21 @@ async function importTransactions(
...
+
                   return {
                     id: entityIdMap.get(subtrans.id),
                     amount: amountFromYnab(subtrans.amount),
                     category: entityIdMap.get(subtrans.category_id) || null,
                     notes: subtrans.memo,
+                    payee: payee,
```
Comment (MatissJanis) [inline packages/loot-core/src/server/importers/ynab5.ts:195]:
```suggestion
                    payee,
```

**Review by MatissJanis — CHANGES_REQUESTED** [review]
Perfect! I was able to test it and it works as expected now. Thanks for clarifying!

Would you mind doing this one small perf improvement? Performance for imports is quite important as some folks come to Actual with years of data (so 100k+ transactions).

File: packages/loot-core/src/server/importers/ynab5.ts (hunk 1 above)
Comment (Marethyu1) [reply packages/loot-core/src/server/importers/ynab5.ts]: Updated based on feedback :) let me know if this isnt quite what you were meaning :)

File: packages/loot-core/src/server/importers/ynab5.ts (hunk 2 above)
Comment (Marethyu1) [reply packages/loot-core/src/server/importers/ynab5.ts]: Added 👍

**Review by MatissJanis — APPROVED** [review]
Perfect! Thanks so much for adding this improvement.
